### PR TITLE
fix(linter): fix performance issues with boundaries lint checks

### PR DESCRIPTION
**Pull request by maxkoretskyi**
## Current Behavior
The circular dependencies check can be very flow in large projects. This is because we traverse the same graph over and over again. Instead of doing this, we can use dynamic programming and cache traversed sub-paths. 

## Expected Behavior
The number of total traversals should become O(EdgesInProjectGraph).

Fixes #
#3393

### CODE REVIEW

File: packages/workspace/src/utils/runtime-lint-utils.spec.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
+import { checkCircularPath } from './runtime-lint-utils';
+
+function transformGraph(graph) {
```
Comment (vsavkin) [inline packages/workspace/src/utils/runtime-lint-utils.spec.ts:3]: should be typed and moved to the bottom of the file

File: packages/workspace/src/utils/runtime-lint-utils.spec.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    };
+  }, {});
+
+  return {
+    nodes,
+    dependencies
+  };
+}
+
+function getPath(graph, node) {
```
Comment (vsavkin) [inline packages/workspace/src/utils/runtime-lint-utils.spec.ts:28]: should be typed and moved to the bottom of the file

File: packages/workspace/src/utils/runtime-lint-utils.spec.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+describe('should find the path between nodes', () => {
+  it('should return empty path when when there are no connecting edges', () => {
+    /*
+
+    A -> B -> C - > E
+
+     */
+
+    const graph = {
+      nodes: ['A', 'B', 'C', 'E'],
```
Comment (vsavkin) [inline packages/workspace/src/utils/runtime-lint-utils.spec.ts:41]: I think you can probably make it a graph with 2 nodes and 1 edge. Based on the assertion, C & E play no role.

File: packages/workspace/src/utils/runtime-lint-utils.spec.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    const g = transformGraph(graph);
+    const path = getPath(g, { from: 'B', to: 'A' });
+
+    expect(path).toEqual([]);
+  });
+
+  it('should find direct path', () => {
+    /*
+
+    A -> B -> C - > E
```
Comment (vsavkin) [inline packages/workspace/src/utils/runtime-lint-utils.spec.ts:58]: There is no E in the graph below

File: packages/workspace/src/utils/runtime-lint-utils.spec.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+    */
+
+    const graph = {
+      nodes: ['A', 'B', 'C'],
```
Comment (vsavkin) [inline packages/workspace/src/utils/runtime-lint-utils.spec.ts:63]: Same. I don't think C is needed, unless you think it changes the test case in a meaningful way

**Review by vsavkin — CHANGES_REQUESTED** [review]
Left some feedback regarding the impl. Nothing major.

1 extra item:
Your unit tests failed because the error message change. You need to update them. But if you look, the error message is weird. It starts with -> and it shoudl not. Maybe add spaces between ->. And I'd make it such that first and the last node in the chain of -> is the same, so it's clear it's a cycle.

**Comment by maxkoretskyi** [comment]
One thing that probably will need some refactoring in the future is unit tests assertions. Currently, an assertion expects certain order of nodes in the path, e.g. for a graph:

```
    A
 /    \
B      C
  \   /
    D

```

and path from `A` to `D` both `[A, B, D]` and `[A, C, D]` are correct results. If we define edges for `A` as:

```
A: [B, C]
```

the function `getPath` will follow `B` node first and return `[A, B, D]`. If we swap `B`, and `C`, the function will return `[A, C, D]`. Although both are correct, the test assertion only expects one correct result. 

As an idea, we could probably introduce our own operator to Jest to check for multiple results:

```
expect(path).toEqualOneOf(['A', 'B', 'D'], ['A', 'C', 'D']);
```

File: packages/workspace/src/tslint/nxEnforceModuleBoundariesRule.spec.ts [hunk 6]
```diff
@@ -796,7 +796,7 @@ describe('Enforce Module Boundaries', () => {
       }
     );
     expect(failures[0].getFailure()).toEqual(
-      'Circular dependency between "mylibName" and "badcirclelibName" detected'
+      'Circular dependency between "mylibName" and "badcirclelibName" detected. The path is: mylibName -> badcirclelibName -> anotherlibName -> mylibName'
```
Comment (vsavkin) [inline packages/workspace/src/tslint/nxEnforceModuleBoundariesRule.spec.ts:799]: How about one of these two messages:

Change the message to ... detected: myLibName -> badcirclarLib...

or

Change the message to ... detected. Path: myLibName -> badcirclarLib...

You don't normally put a colon after "is".

**Comment by vsavkin** [comment]
merged here https://github.com/nrwl/nx/pull/3824

**Pull request closed without merging**